enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,288 @@
+3. Write RHCOS image and the master Ignition to disk.
+4. Reboot the node.
+
+#### Cluster-bootstrap
+
+By default, `cluster-bootstrap` start the bootstrap control plane and create all the manifests under /opt/openshift/manifests.
+`cluster-bootstrap` also wait for a list required pods to be ready, these pods are expetced to start running on the master nodes.
+In case we are running the bootstrap in place, there is no master node that can run the pods. cluster-bootstrap should apply the manifest and tear down the control plane, in case it fail to apply some of the manifest it should return an error.
+
+
+`cluster-bootstrap` will have a new entrypoint `--bootstrap-in-place` which will get the master Ignition as input and will enrich the master Ignition with control plane static pods manifests and all required resources including etcd data.
+
+#### Bootstrap / Control plane static pods
+The control plane components we will add to the master Ignition are (to be placed under `/etc/kubernetes/manifests`):
+1. etcd-pod
+2. kube-apiserver-pod
+3. kube-controller-manager-pod
+4. kube-scheduler-pod
+
+Control plane required resources to be added to the Ignition:
+1. `/var/lib/etcd`
+2. `/etc/kubernetes/bootstrap-configs`
+3. /opt/openshift/tls/* (`/etc/kubernetes/bootstrap-secrets`)
+4. /opt/openshift/auth/kubeconfig-loopback (`/etc/kubernetes/bootstrap-secrets/kubeconfig`)
+
+Note: `/etc/kubernetes/bootstrap-secrets` and `/etc/kubernetes/bootstrap-configs` will be deleted post reboot once the OCP control plane is ready.
+
+The control plane opertors (that will run on the node post reboot) will manage the rollout of new revision of the control plane pods
+
+#### Etcd data
+
+In order to add a viable, working etcd post reboot, we will take a snapshot of etcd and add it to the master Ignition.
+Post reboot we will use restore etcd-member from the snapshot.
+
+Another option is to stop the etcd pod (move the static pod manifest from `/etc/kubernetes/manifests`).
+Let etcd save its state and exit. once etcd is down. we will add the `/var/lib/etcd` directory to the master Ignition config.
+After the reboot etcd should start with all the data it had prior to the reboot.
+
+#### Post reboot
+
+We will add a new post-reboot service for approving the kubelet and the node Certificate Sign Requests.
+This service will also cleanup the bootstrap static pods resources once the OCP control plane is ready.
+Since we start with a liveCD, the bootstrap services (bootkube, approve-csr, etc), /etc and /opt/openshift "garbage" is left behind and doesn't get to the node (post reboot).
+The files that we need to delete are under:
+`/etc/kubernetes/bootstrap-secrets` and `/etc/kubernetes/bootstrap-configs`
+These files are required for the bootstrap control plane to start before it's replaced by the control plane operators.
+Once the OCP control plane static pods are deployed we can delete the files as they are no longer required. 
```

Comment:
Do we preserve bootstrap logs anywhere? How do we debug this when it goes wrong either prior to the reboot or after?

Reply:
You are correct, debugging it when things go wrong is a challenge since there is no bootstrap node with all the logs.
We can run installer-gather prior to the reboot but there is no place to keep it.
I think that if we provide a way to pause the installation before the reboot happens (unsure what will be the API for that), it should allow the user to login and run installer-gather.
Debugging issues post node reboot will be done by ssh  or oc.

Reply:
1. we can send them to the hub cluster in case the installation will be orchestrated by the assisted installer
2. maybe, we can consider pushing them via /boot partition into the master node?

Reply:
A big +1 to ensuring that debugging capabilities are retained for b-i-p. 

During bootstrap and before reboot, will ssh access to the node be possible if something goes wrong?

Reply:
once we allow to stop before the reboot then yes

Reply:
Can we just copy the contents of `/var/log/journal` to the new system? Is that enough context? The admin could then use `journalctl --machine` to select the logs from the live CD.

Reply:
If something goes wrong during bootstrap we will have all debugging capabilities, just like regular installation.
Once bootstrapping has completed, the bootstrap node is torn down, and the journals are lost, same as In a regular installation flow.
I don't think it should be different here.

Reply:
I agree with that, but there is a corner case: if bootstrapping completes but the cluster installation fails. This bootstrap-in-place flow tears down the bootstrap node much earlier than our standard IPI flow, right? If an operator has been misconfigured such that the bootstrap completes but the Cluster Version never reports a successful install, the user might not be able to inspect the logs that led up to the failure. This might be too contrived a use case, and we might decide that it's okay to ignore this for now, but I still want to call it out.

Reply:
@eranco74 What do you mean that the 'bootstrap is torn down'? Isn't it the same node? What would preclude the install saving bootstrap artifacts?

Reply:
"Bootstrap" in @eranco74's statement refers to the logical entity; not the physical machine. Tearing down the bootstrap node, in practice, means leaving the live environment (ephemeral) and rebooting into the installed system. In effect, it's a different logical machine (e.g. `/etc/machine-id` is different) and there is no shared state.

Reply:
@crawford you are correct, unlike 3 nodes installation for bootstrap in place we lose the bootstrap node before we have a running control plane.
I'll add a section for preserve bootstrap logs when we pivot the node and allowing to debug the the installation using installer-gather prior to the reboot and after.